### ISSUE TITLE
[Arm64/Unix] CLREX unrecognized by Valgrind 3.12 

### ISSUE THREAD

**Issue by sdmaclea** [dotnet/coreclr#10612]
There was a defect against valgrind.  [Bug 359838](https://bugs.kde.org/show_bug.cgi?id=359838) - arm64: Unhandled instruction 0xD5033F5F (clrex).  This defect is cited as being fixed in the release notes for 3.12.  http://valgrind.org/docs/manual/dist.news.html.

CoreCLR always hits the defect even in release 3.12.

```
==10212== Memcheck, a memory error detector
==10212== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==10212== Using Valgrind-3.12.0.SVN and LibVEX; rerun with -h for copyright info
==10212== Command: /home/sdmaclea/overlay.vmj/corerun /home/sdmaclea/hwapp/bin/Debug/netcoreapp1.1/hwapp.dll
==10212== 
==10212== Warning: set address range perms: large range [0x393b8000, 0xb2fa8000) (noaccess)
==10212== Warning: set address range perms: large range [0x393b8000, 0xb2fa8000) (noaccess)
==10212== Warning: set address range perms: large range [0x393b8000, 0xaafa8000) (noaccess)
==10212== Warning: set address range perms: large range [0x393b8000, 0xaafa8000) (noaccess)
==10212== Warning: set address range perms: large range [0x393b8000, 0xa2fa8000) (noaccess)
==10212== Warning: set address range perms: large range [0x393b8000, 0xa2fa8000) (noaccess)
==10212== Warning```
: set address range perms: large range [0x393b8000, 0x9afa8000) (noaccess)
==10212== Warning: set address range perms: large range [0x393b8000, 0x9afa8000) (noaccess)
==10212== Warning: set address range perms: large range [0x393b8000, 0x92fa8000) (noaccess)
==10212== Warning: set address range perms: large range [0x393b8000, 0x92fa8000) (noaccess)
==10212== Warning: set address range perms: large range [0x393b8000, 0x8afa8000) (noaccess)
==10212== Warning: set address range perms: large range [0x393b8000, 0x8afa8000) (noaccess)
==10212== Warning: set address range perms: large range [0x3b454000, 0x85044000) (noaccess)
ARM64 front end: branch_etc
disInstr(arm64): unhandled instruction 0xD5033F5F
disInstr(arm64): 1101'0101 0000'0011 0011'1111 0101'1111
==10212== valgrind: Unrecognised instruction at address 0x51861cc.
==10212==    at 0x51861CC: InterlockedCompareExchange (pal.h:4361)
==10212==    by 0x59DEB33: CorUnix::CriticalSectionSubSysInitialize() (cs.cpp:568)
==10212==    by 0x597A133: Initialize(int, char const* const*, unsigned int) (pal.cpp:213)
==10212==    by 0x597B5A7: PAL_InitializeDLL (pal.cpp:168)
==10212==    by 0x5982B6F: PAL_RegisterModule (module.cpp:679)
==10212==    by 0x59829D7: LOADRegisterLibraryDirect(void*, char const*, int) (module.cpp:1559)
==10212==    by 0x597FE1F: LOADLoadLibrary(char const*, int) (module.cpp:1637)
==10212==    by 0x5984AA3: LOADGetPalLibrary (module.cpp:1722)
==10212==    by 0x5984697: LOADInitializeCoreCLRModule (module.cpp:1659)
==10212==    by 0x597B637: PAL_InitializeCoreCLR (pal.cpp:648)
==10212==    by 0x515827F: coreclr_initialize (unixinterface.cpp:175)
==10212==    by 0x405F93: ExecuteManagedAssembly(char const*, char const*, char const*, int, char const**) (coreruncommon.cpp:404)
==10212== Your program just tried to execute an instruction that Valgrind
==10212== did not recognise.  There are two possible reasons for this.
==10212== 1. Your program has a bug and erroneously jumped to a non-code
==10212==    location.  If you are running Memcheck and you just saw a
==10212==    warning about a bad jump, it's probably your program's fault.
==10212== 2. The instruction is legitimate but Valgrind doesn't handle it,
==10212==    i.e. it's Valgrind's fault.  If you think this is the case or
==10212==    you are not sure, please let us know and we'll try to fix it.
==10212== Either way, Valgrind will now raise a SIGILL signal which will
==10212== probably kill your program.
==10212== valgrind: Unrecognised instruction at address 0x51861cc.
==10212==    at 0x51861CC: InterlockedCompareExchange (pal.h:4361)
==10212==    by 0x59DEB33: CorUnix::CriticalSectionSubSysInitialize() (cs.cpp:568)
==10212==    by 0x597A133: Initialize(int, char const* const*, unsigned int) (pal.cpp:213)
==10212==    by 0x597B5A7: PAL_InitializeDLL (pal.cpp:168)
==10212==    by 0x5982B6F: PAL_RegisterModule (module.cpp:679)
==10212==    by 0x59829D7: LOADRegisterLibraryDirect(void*, char const*, int) (module.cpp:1559)
==10212==    by 0x597FE1F: LOADLoadLibrary(char const*, int) (module.cpp:1637)
==10212==    by 0x5984AA3: LOADGetPalLibrary (module.cpp:1722)
==10212==    by 0x5984697: LOADInitializeCoreCLRModule (module.cpp:1659)
==10212==    by 0x597B637: PAL_InitializeCoreCLR (pal.cpp:648)
==10212==    by 0x515827F: coreclr_initialize (unixinterface.cpp:175)
==10212==    by 0x405F93: ExecuteManagedAssembly(char const*, char const*, char const*, int, char const**) (coreruncommon.cpp:404)
==10212== Your program just tried to execute an instruction that Valgrind
==10212== did not recognise.  There are two possible reasons for this.
==10212== 1. Your program has a bug and erroneously jumped to a non-code
==10212==    location.  If you are running Memcheck and you just saw a
==10212==    warning about a bad jump, it's probably your program's fault.
==10212== 2. The instruction is legitimate but Valgrind doesn't handle it,
==10212==    i.e. it's Valgrind's fault.  If you think this is the case or
==10212==    you are not sure, please let us know and we'll try to fix it.
==10212== Either way, Valgrind will now raise a SIGILL signal which will
```

**Comment by sdmaclea** [comment]
@janvorli FYI

**Comment by sdmaclea** [comment]
clrex is coming from the use of the gcc __sync builtins

**Comment by sdmaclea** [comment]
The valgrind from debian sid (unstable) has the clrex fix (valgrind_3.12.0-1.1_arm64.deb).  The valgrind 3.12 in debian stretch (tested above) is currently a pre-released valgrind 3.12 version which does not include the fix above.
